### PR TITLE
BUG/REG: file-handle object handled incorrectly in to_csv

### PR DESCRIPTION
**Pull request by minggli**
- [x] closes #https://github.com/pandas-dev/pandas/issues/21471
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This error related to PR #21249 and https://github.com/pandas-dev/pandas/issues/21227. This is never supported use case and to use file-handle in to_csv with compression, the file-object itself should be a compression archive such as:
```
with gzip.open('test.txt.gz', 'wt') as f:
    pd.DataFrame([0,1],index=['a','b'], columns=['c']).to_csv(f, sep='\t')
```

Regressed to 0.22 to_csv with support for zipfile. zipfile doesn't support writing csv strings to a zip archive using a file-handle. So buffer is used to catch the writing and dump into zip archive in one go. The other scenarios remain unchanged.

### CODE REVIEW

**Comment by wildraid** [comment]
I am not a proper reviewer, but I am suspicious about the whole "ZipFile magic" part.

It looks really ugly to me, and this code may potentially cause spike of 2x memory usage due to StringIO accumulation and "getvalue()" call.

I do not fully understand what was the problem with 0.22 implementation. Yes, "read_csv" and "to_csv" were inconsistent, but it contained less "magic", and users could always wrap anything they need with few lines of code.

Also, compression algos usually have additional parameters. Users might want to change defaults later on, and they'll need manual wrapping for that anyway.

Just thoughts.

**Comment by minggli** [comment]
agree that for extremely large zip output, memory usage will increase when writing csv to StringIO first and then call `getvalue()`. With 8G RAM MacOS, enough to write a dataframe [100m, 3]. the memory increase is only when zip compression, for other use cases, there should be no difference as 0.22. the custom class just provides a file-like that accepts string into zip archive when ZipFile class doesn't really offer that. 

The ability to write zip archive is added https://github.com/pandas-dev/pandas/issues/17778 where only read zip was supported, whereas we have round-trip ability for others. the motivation is consistency but also zip compression format is quite common and strange not to be supported.

the zip class is not only for to_csv but provides utility for to_json and to_pickle case too.

File: doc/source/whatsnew/v0.23.2.txt [hunk 1]
```diff
@@ -56,7 +56,7 @@ Bug Fixes
 
 **I/O**
 
--
+- Bug in :meth:`to_csv` when handling file-like object incorrectly (:issue:`21471`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.2.txt]: can you move to the regression section

File: pandas/io/formats/csvs.py [hunk 2]
```diff
@@ -128,19 +128,33 @@ def save(self):
...
-        # file-like output with compression as option. GH 21241, 21118
-        f = StringIO()
-        if not is_file_like(self.path_or_buf):
-            # path_or_buf is path
-            path_or_buf = self.path_or_buf
-        elif hasattr(self.path_or_buf, 'name'):
-            # path_or_buf is file handle
-            path_or_buf = self.path_or_buf.name
+        if self.compression and hasattr(self.path_or_buf, 'write'):
+            import warnings
```
Comment (jreback) [inline pandas/io/formats/csvs.py]: you can move this import to the top

File: pandas/io/formats/csvs.py [hunk 3]
```diff
@@ -128,19 +128,33 @@ def save(self):
...
-        # PR 21300 uses string buffer to receive csv writing and dump into
-        # file-like output with compression as option. GH 21241, 21118
-        f = StringIO()
-        if not is_file_like(self.path_or_buf):
-            # path_or_buf is path
-            path_or_buf = self.path_or_buf
-        elif hasattr(self.path_or_buf, 'name'):
-            # path_or_buf is file handle
-            path_or_buf = self.path_or_buf.name
+        if self.compression and hasattr(self.path_or_buf, 'write'):
```
Comment (jreback) [inline pandas/io/formats/csvs.py]: can uou add a comment here (sure the warning says it all, but helpful nonetheless), also an issue reference

File: pandas/io/formats/csvs.py [hunk 4]
```diff
@@ -128,19 +128,33 @@ def save(self):
...
+            import warnings
+            msg = ("compression has no effect when passing file-like "
+                   "object as input.")
+            warnings.warn(msg, RuntimeWarning, stacklevel=2)
+
+        if isinstance(self.path_or_buf, ZipFile) or (
```
Comment (jreback) [inline pandas/io/formats/csvs.py]: you can just do

``is_zip = isinstance(......) or (.....)``

Comment (jreback) [inline pandas/io/formats/csvs.py]: comment here on what is_zip means

File: pandas/io/formats/csvs.py [hunk 5]
```diff
@@ -157,13 +171,18 @@ def save(self):
...
-                                         encoding=encoding,
-                                         compression=self.compression)
-                f.write(buf)
+            if is_zip:
+                # GH 17778 handles zip compression separately.
+                buf = f.getvalue()
+                try:
+                    self.path_or_buf.write(buf)
+                except AttributeError:
+                    f, handles = _get_handle(self.path_or_buf, self.mode,
```
Comment (jreback) [inline pandas/io/formats/csvs.py]: can you add a comment on when the except happens?

File: pandas/tests/series/test_io.py [hunk 6]
```diff
@@ -151,19 +152,19 @@ def test_to_csv_compression(self, s, encoding, compression):
...
-            with open(filename, 'w') as fh:
-                s.to_csv(fh, compression=compression, encoding=encoding,
-                         header=True)
-
+            # test the round trip using file handle - to_csv -> read_csv
+            f, _handles = _get_handle(filename, 'w', compression=compression,
+                                      encoding=encoding)
+            with f:
+                s.to_csv(f, encoding=encoding, header=True)
             result_fh = pd.read_csv(filename, compression=compression,
```
Comment (jreback) [inline pandas/tests/series/test_io.py]: you can rename this to result (and below)

File: pandas/tests/test_common.py [hunk 7]
```diff
@@ -246,19 +247,31 @@ def test_compression_size(obj, method, compression_only):
...
+        f, _handles = _get_handle(filename, 'w', compression=None)
+        with f:
+            getattr(obj, method)(f)
+            assert not f.closed
         uncompressed = os.path.getsize(filename)
         assert uncompressed > compressed
+
+
+def test_compression_warning(compression_only):
+    df = DataFrame(100 * [[0.123456, 0.234567, 0.567567],
```
Comment (jreback) [inline pandas/tests/test_common.py]: can you add the issue number

**Review by jreback — CHANGES_REQUESTED** [review]
looks good. mostly doc comments. ping on green.

File: pandas/tests/test_common.py (hunk 7 above)
Comment (minggli) [reply pandas/tests/test_common.py]: added.

File: pandas/io/formats/csvs.py (hunk 5 above)
Comment (minggli) [reply pandas/io/formats/csvs.py]: done.

File: pandas/tests/series/test_io.py (hunk 6 above)
Comment (minggli) [reply pandas/tests/series/test_io.py]: done.

File: doc/source/whatsnew/v0.23.2.txt (hunk 1 above)
Comment (minggli) [reply doc/source/whatsnew/v0.23.2.txt]: done.

File: pandas/io/formats/csvs.py [hunk 8]
```diff
@@ -157,13 +171,19 @@ def save(self):
...
-            if path_or_buf:
-                f, handles = _get_handle(path_or_buf, self.mode,
-                                         encoding=encoding,
-                                         compression=self.compression)
-                f.write(buf)
+            if is_zip:
+                # GH 17778 handles zip compression separately.
+                buf = f.getvalue()
+                try:
+                    self.path_or_buf.write(buf)
```
Comment (gfyoung) [inline pandas/io/formats/csvs.py]: They say that it's sometimes easier to ask for forgiveness than it is for permission.

That being said, your treatment of `self.path_or_buf.write` is inconsistent here compared to:

https://github.com/pandas-dev/pandas/pull/21478/files#diff-eaa887c826bfb361d98db1ddb668c7deR150

In addition, the logic seems somewhat similar.  Can we potentially deduplicate this logic?

File: doc/source/whatsnew/v0.23.2.txt [hunk 9]
```diff
@@ -16,7 +16,7 @@ and bug fixes. We recommend that all users upgrade to this version.
 Fixed Regressions
 ~~~~~~~~~~~~~~~~~
 
--
+- Fixed Regression in :meth:`to_csv` when handling file-like object incorrectly (:issue:`21471`)
```
Comment (gfyoung) [inline doc/source/whatsnew/v0.23.2.txt]: nit: "Regression" --> "regression"

File: pandas/io/formats/csvs.py (hunk 8 above)
Comment (minggli) [reply pandas/io/formats/csvs.py]: see your point. changed to if else to make it consistent.

File: doc/source/whatsnew/v0.23.2.txt (hunk 9 above)
Comment (minggli) [reply doc/source/whatsnew/v0.23.2.txt]: done.

**Comment by minggli** [comment]
@jreback comments carried out.

**Review by gfyoung — APPROVED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @minggli nice followup patch!

**Comment by jreback** [comment]
@TomAugspurger this might be tricky to backport

**Comment by minggli** [comment]
@jreback @gfyoung happy to help. thanks for reviewing.

**Pull request merged**